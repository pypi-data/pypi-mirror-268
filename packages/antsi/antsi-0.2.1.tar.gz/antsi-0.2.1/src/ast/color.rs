use std::{
    fmt::{Display, Formatter},
    str::FromStr,
};

/// The provided [`Color`] name was invalid
#[derive(Clone, Copy, Debug)]
pub struct InvalidColorError;

impl std::error::Error for InvalidColorError {}

impl Display for InvalidColorError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.write_str("invalid color name")
    }
}

macro_rules! colors {
    (
        $( $( #[ $meta:meta ] )* $color:ident $fg:literal $bg:literal ( $names:pat ) ),* $(,)?
    ) => {
        /// Available standard ANSI colors
        #[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
        pub enum Color {
            $( $( #[$meta] )* $color, )*
        }

        impl Color {
            /// Convert to the foreground ANSI code
            pub fn foreground_code(&self) -> &'static str {
                match self {
                    $( Color::$color => stringify!($fg), )*
                }
            }

            /// Convert to the background ANSI code
            pub fn background_code(&self) -> &'static str {
                match self {
                    $( Color::$color => stringify!($bg), )*
                }
            }
        }

        impl FromStr for Color {
            type Err = InvalidColorError;

            fn from_str(name: &str) -> Result<Self, Self::Err> {
                Ok(match name.to_ascii_lowercase().as_str() {
                    $( $names => Color::$color, )*
                    _ => return Err(InvalidColorError),
                })
            }
        }
    };
}

colors! {
    Black   30 40 ("black"),
    Red     31 41 ("red"),
    Green   32 42 ("green"),
    Yellow  33 43 ("yellow"),
    Blue    34 44 ("blue"),
    Magenta 35 45 ("magenta"),
    Cyan    36 46 ("cyan"),
    White   37 47 ("white"),
    #[default]
    Default 39 49 ("default"),

    BrightBlack   90 100 ("bright-black"),
    BrightRed     91 101 ("bright-red"),
    BrightGreen   92 102 ("bright-green"),
    BrightYellow  93 103 ("bright-yellow"),
    BrightBlue    94 104 ("bright-blue"),
    BrightMagenta 95 105 ("bright-magenta"),
    BrightCyan    96 106 ("bright-cyan"),
    BrightWhite   97 107 ("bright-white"),
}
