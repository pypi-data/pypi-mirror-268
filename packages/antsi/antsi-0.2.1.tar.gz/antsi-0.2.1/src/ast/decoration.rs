use std::{
    fmt::{Display, Formatter},
    str::FromStr,
};

/// The provided [`Decoration`] name was invalid
#[derive(Clone, Copy, Debug)]
pub struct InvalidDecorationError;

impl std::error::Error for InvalidDecorationError {}

impl Display for InvalidDecorationError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.write_str("invalid decoration name")
    }
}

macro_rules! decorations {
    (
        $( $decoration:ident $apply:literal $remove:literal ( $names:pat ) ),* $(,)?
    ) => {
        /// Available standard ANSI text decorations
        #[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
        pub enum Decoration {
            $( $decoration, )*
        }

        impl Decoration {
            /// Convert to the ANSI code for applying the styling
            pub fn apply_code(&self) -> &'static str {
                match self {
                    $( Decoration::$decoration => stringify!($apply), )*
                }
            }

            /// Convert to the ANSI code for removing the styling
            pub fn remove_code(&self) -> &'static str {
                match self {
                    $( Decoration::$decoration => stringify!($remove), )*
                }
            }
        }

        impl FromStr for Decoration {
            type Err = InvalidDecorationError;

            fn from_str(name: &str) -> Result<Self, Self::Err> {
                Ok(match name.to_ascii_lowercase().as_str() {
                    $( $names => Decoration::$decoration, )*
                    _ => return Err(InvalidDecorationError),
                })
            }
        }
    };
}

decorations! {
    Bold          1 22 ("bold"),
    Dim           2 22 ("dim" | "faint"),
    Italic        3 23 ("italic"),
    Underline     4 24 ("underline"),
    SlowBlink     5 25 ("slow-blink" | "blink-slow"),
    FastBlink     6 25 ("fast-blink" | "blink-fast"),
    Invert        7 27 ("invert" | "reverse"),
    Hide          8 28 ("hide" | "conceal"),
    StrikeThrough 9 29 ("strikethrough" | "strike-through"),
}
