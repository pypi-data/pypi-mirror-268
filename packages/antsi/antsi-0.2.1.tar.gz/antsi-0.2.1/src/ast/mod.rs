mod color;
mod decoration;
mod style;
mod token;

#[allow(unused_imports)]
pub use color::{Color, InvalidColorError};
#[allow(unused_imports)]
pub use decoration::{Decoration, InvalidDecorationError};
pub use style::{CurrentStyle, Style};
pub use token::{Token, Tokens};
