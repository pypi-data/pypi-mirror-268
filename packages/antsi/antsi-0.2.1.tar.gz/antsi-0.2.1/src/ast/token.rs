use super::Style;

/// One or more pieces of text that either apply additional styling or inherit from the parent
/// styles.
#[derive(Clone, Debug)]
#[cfg_attr(test, derive(Eq, PartialEq))]
pub enum Token {
    /// A piece of text that does not modify the styling
    Content(String),
    /// One or more pieces of text that with additional styling
    Styled {
        /// The pieces of text the style applies to
        content: Vec<Token>,
        /// The style to apply
        style: Style,
    },
}

/// A sequence of [`Token`]s
#[derive(Clone, Debug, Default)]
#[cfg_attr(test, derive(Eq, PartialEq))]
pub struct Tokens(Vec<Token>);

impl From<Vec<Token>> for Tokens {
    fn from(tokens: Vec<Token>) -> Self {
        Tokens(tokens)
    }
}

impl From<Tokens> for Vec<Token> {
    fn from(tokens: Tokens) -> Self {
        tokens.0
    }
}

impl Tokens {
    /// Add a new token to the end of the sequence
    pub fn push(&mut self, token: Token) {
        self.0.push(token)
    }

    /// Add a string to the end of the sequence
    ///
    /// If the last token in the sequence is an unstyled piece of text, it will be appended directly
    /// to the token. Otherwise, a new content token will be created.
    pub fn push_str(&mut self, s: &str) {
        match self.0.last_mut() {
            Some(Token::Content(content)) => content.push_str(s),
            Some(Token::Styled { .. }) | None => self.0.push(Token::Content(s.to_string())),
        }
    }

    /// Add a character to the end of the sequence
    ///
    /// If the last token in the sequence is an unstyled piece of text, it will be appended directly
    /// to the token. Otherwise, a new content token will be created.
    pub fn push_char(&mut self, ch: char) {
        match self.0.last_mut() {
            Some(Token::Content(content)) => content.push(ch),
            Some(Token::Styled { .. }) | None => self.0.push(Token::Content(ch.to_string())),
        }
    }

    /// Add all the tokens from another sequence onto the current one
    pub fn extend<T>(&mut self, other: T)
    where
        T: Into<Tokens>,
    {
        let other = other.into();
        if other.0.is_empty() {
            return;
        } else if self.0.is_empty() {
            *self = other;
            return;
        }

        let mut tokens = other.0.into_iter().peekable();
        if let Some(Token::Content(last)) = self.0.last_mut() {
            while let Some(Token::Content(content)) = tokens.peek() {
                last.push_str(content);
                tokens.next();
            }
        }

        self.0.extend(tokens);
    }
}

#[cfg(test)]
mod tests {
    use super::{Token, Tokens};

    #[test]
    fn push_adds_token_to_end_when_no_tokens() {
        let mut tokens = Tokens::default();
        tokens.push(Token::Content(String::from("test")));

        assert_eq!(
            tokens,
            Tokens::from(vec![Token::Content(String::from("test"))])
        )
    }

    #[test]
    fn push_adds_token_to_end_when_tokens_present() {
        let mut tokens = Tokens::from(vec![Token::Content(String::from("existing"))]);
        tokens.push(Token::Content(String::from("test")));

        assert_eq!(
            tokens,
            Tokens::from(vec![
                Token::Content(String::from("existing")),
                Token::Content(String::from("test"))
            ])
        )
    }

    #[test]
    fn push_str_adds_new_content_token_to_end_when_no_tokens() {
        let mut tokens = Tokens::default();
        tokens.push_str("test");

        assert_eq!(
            tokens,
            Tokens::from(vec![Token::Content(String::from("test"))])
        )
    }

    #[test]
    fn push_str_adds_new_content_token_to_end_when_empty_style_token_present() {
        let mut tokens = Tokens::from(vec![Token::Styled {
            content: vec![],
            style: style!(),
        }]);
        tokens.push_str("test");

        assert_eq!(
            tokens,
            Tokens::from(vec![
                Token::Styled {
                    content: vec![],
                    style: style!()
                },
                Token::Content(String::from("test"))
            ])
        )
    }

    #[test]
    fn push_str_adds_new_content_token_to_end_when_style_token_with_children_exists() {
        let mut tokens = Tokens::from(vec![Token::Styled {
            content: vec![Token::Content(String::from("existing"))],
            style: style!(),
        }]);
        tokens.push_str("test");

        assert_eq!(
            tokens,
            Tokens::from(vec![
                Token::Styled {
                    content: vec![Token::Content(String::from("existing"))],
                    style: style!(),
                },
                Token::Content(String::from("test"))
            ])
        );
    }

    #[test]
    fn push_str_appends_to_last_non_nested_content_token() {
        let mut tokens = Tokens::from(vec![Token::Content(String::from("existing "))]);
        tokens.push_str("test");

        assert_eq!(
            tokens,
            Tokens::from(vec![Token::Content(String::from("existing test"))])
        );
    }

    #[test]
    fn push_char_adds_new_content_token_to_end_when_no_tokens() {
        let mut tokens = Tokens::default();
        tokens.push_char('T');

        assert_eq!(
            tokens,
            Tokens::from(vec![Token::Content(String::from("T"))])
        )
    }

    #[test]
    fn push_char_adds_new_content_token_to_end_when_empty_style_token_present() {
        let mut tokens = Tokens::from(vec![Token::Styled {
            content: vec![],
            style: style!(),
        }]);
        tokens.push_char('T');

        assert_eq!(
            tokens,
            Tokens::from(vec![
                Token::Styled {
                    content: vec![],
                    style: style!()
                },
                Token::Content(String::from("T"))
            ])
        )
    }

    #[test]
    fn push_char_adds_new_content_token_to_end_when_style_token_with_children_exists() {
        let mut tokens = Tokens::from(vec![Token::Styled {
            content: vec![Token::Content(String::from("existing"))],
            style: style!(),
        }]);
        tokens.push_char('T');

        assert_eq!(
            tokens,
            Tokens::from(vec![
                Token::Styled {
                    content: vec![Token::Content(String::from("existing"))],
                    style: style!(),
                },
                Token::Content(String::from("T"))
            ])
        );
    }

    #[test]
    fn push_char_appends_to_last_non_nested_content_token() {
        let mut tokens = Tokens::from(vec![Token::Content(String::from("existing "))]);
        tokens.push_char('T');

        assert_eq!(
            tokens,
            Tokens::from(vec![Token::Content(String::from("existing T"))])
        );
    }

    #[test]
    fn extend_both_empty() {
        let mut tokens = Tokens::default();
        tokens.extend(Tokens::default());
        assert_eq!(tokens, Tokens::from(vec![]));
    }

    #[test]
    fn extend_empty_with_content_token() {
        let mut tokens = Tokens::default();
        tokens.extend(vec![Token::Content(String::from("testing"))]);
        assert_eq!(
            tokens,
            Tokens::from(vec![Token::Content(String::from("testing"))])
        );
    }

    #[test]
    fn extend_empty_with_styled_token() {
        let mut tokens = Tokens::default();
        tokens.extend(vec![Token::Styled {
            style: style!(),
            content: vec![Token::Content(String::from("testing"))],
        }]);
        assert_eq!(
            tokens,
            Tokens::from(vec![Token::Styled {
                style: style!(),
                content: vec![Token::Content(String::from("testing"))]
            }])
        );
    }

    #[test]
    fn extend_single_content_token_with_empty() {
        let mut tokens = Tokens::from(vec![Token::Content(String::from("existing"))]);
        tokens.extend(vec![]);
        assert_eq!(
            tokens,
            Tokens::from(vec![Token::Content(String::from("existing"))])
        );
    }

    #[test]
    fn extend_single_content_token_with_content_token() {
        let mut tokens = Tokens::from(vec![Token::Content(String::from("existing"))]);
        tokens.extend(vec![Token::Content(String::from("testing"))]);
        assert_eq!(
            tokens,
            Tokens::from(vec![Token::Content(String::from("existingtesting"))])
        );
    }

    #[test]
    fn extend_single_content_token_with_styled_token() {
        let mut tokens = Tokens::from(vec![Token::Content(String::from("existing"))]);
        tokens.extend(vec![Token::Styled {
            style: style!(),
            content: vec![Token::Content(String::from("testing"))],
        }]);
        assert_eq!(
            tokens,
            Tokens::from(vec![
                Token::Content(String::from("existing")),
                Token::Styled {
                    style: style!(),
                    content: vec![Token::Content(String::from("testing"))],
                }
            ])
        );
    }

    #[test]
    fn extend_single_styled_token_with_empty() {
        let mut tokens = Tokens::from(vec![Token::Styled {
            style: style!(),
            content: vec![Token::Content(String::from("testing"))],
        }]);
        tokens.extend(vec![]);
        assert_eq!(
            tokens,
            Tokens::from(vec![Token::Styled {
                style: style!(),
                content: vec![Token::Content(String::from("testing"))],
            }])
        );
    }

    #[test]
    fn extend_single_styled_token_with_content_token() {
        let mut tokens = Tokens::from(vec![Token::Styled {
            style: style!(),
            content: vec![Token::Content(String::from("testing"))],
        }]);
        tokens.extend(vec![Token::Content(String::from("testing"))]);
        assert_eq!(
            tokens,
            Tokens::from(vec![
                Token::Styled {
                    style: style!(),
                    content: vec![Token::Content(String::from("testing"))],
                },
                Token::Content(String::from("testing"))
            ])
        );
    }

    #[test]
    fn extend_single_styled_token_with_styled_token() {
        let mut tokens = Tokens::from(vec![Token::Styled {
            style: style!(),
            content: vec![Token::Content(String::from("testing"))],
        }]);
        tokens.extend(vec![Token::Styled {
            style: style!(),
            content: vec![Token::Content(String::from("testing"))],
        }]);
        assert_eq!(
            tokens,
            Tokens::from(vec![
                Token::Styled {
                    style: style!(),
                    content: vec![Token::Content(String::from("testing"))],
                },
                Token::Styled {
                    style: style!(),
                    content: vec![Token::Content(String::from("testing"))],
                }
            ])
        );
    }

    #[test]
    fn extend_styled_and_content_token_with_styled_token() {
        let mut tokens = Tokens::from(vec![
            Token::Styled {
                style: style!(),
                content: vec![Token::Content(String::from("existing styled"))],
            },
            Token::Content(String::from("existing content")),
        ]);
        tokens.extend(vec![Token::Styled {
            style: style!(),
            content: vec![Token::Content(String::from("testing"))],
        }]);
        assert_eq!(
            tokens,
            Tokens::from(vec![
                Token::Styled {
                    style: style!(),
                    content: vec![Token::Content(String::from("existing styled"))],
                },
                Token::Content(String::from("existing content")),
                Token::Styled {
                    style: style!(),
                    content: vec![Token::Content(String::from("testing"))],
                }
            ])
        );
    }

    #[test]
    fn extend_content_and_styled_token_with_styled_token() {
        let mut tokens = Tokens::from(vec![
            Token::Content(String::from("existing content")),
            Token::Styled {
                style: style!(),
                content: vec![Token::Content(String::from("existing styled"))],
            },
        ]);
        tokens.extend(vec![Token::Styled {
            style: style!(),
            content: vec![Token::Content(String::from("testing"))],
        }]);
        assert_eq!(
            tokens,
            Tokens::from(vec![
                Token::Content(String::from("existing content")),
                Token::Styled {
                    style: style!(),
                    content: vec![Token::Content(String::from("existing styled"))],
                },
                Token::Styled {
                    style: style!(),
                    content: vec![Token::Content(String::from("testing"))],
                }
            ])
        );
    }

    #[test]
    fn extend_styled_and_content_token_with_content_token() {
        let mut tokens = Tokens::from(vec![
            Token::Styled {
                style: style!(),
                content: vec![Token::Content(String::from("existing styled"))],
            },
            Token::Content(String::from("existing content")),
        ]);
        tokens.extend(vec![Token::Content(String::from("testing"))]);
        assert_eq!(
            tokens,
            Tokens::from(vec![
                Token::Styled {
                    style: style!(),
                    content: vec![Token::Content(String::from("existing styled"))],
                },
                Token::Content(String::from("existing contenttesting")),
            ])
        );
    }

    #[test]
    fn extend_content_and_styled_token_with_content_token() {
        let mut tokens = Tokens::from(vec![
            Token::Content(String::from("existing content")),
            Token::Styled {
                style: style!(),
                content: vec![Token::Content(String::from("existing styled"))],
            },
        ]);
        tokens.extend(vec![Token::Content(String::from("testing"))]);
        assert_eq!(
            tokens,
            Tokens::from(vec![
                Token::Content(String::from("existing content")),
                Token::Styled {
                    style: style!(),
                    content: vec![Token::Content(String::from("existing styled"))],
                },
                Token::Content(String::from("testing"))
            ])
        );
    }
}
