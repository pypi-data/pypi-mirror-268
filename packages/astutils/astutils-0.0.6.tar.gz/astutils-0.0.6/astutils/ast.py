"""Abstract syntax tree nodes."""
# Copyright 2014-2022 by California Institute of Technology
# All rights reserved. Licensed under 3-clause BSD.
#


class Terminal:
    """Nullary symbol."""

    def __init__(
            self,
            value:
                str,
            dtype:
                str='terminal'
            ) -> None:
        try:
            value + 's'
        except TypeError:
            raise TypeError(
                'value must be a string, '
                f'got: {value}')
        self.type = dtype
        self.value = value

    def __hash__(
            self
            ) -> int:
        return id(self)

    def __repr__(
            self
            ) -> str:
        class_name = type(self).__name__
        return (
            f'{class_name}('
                f'{self.value!r}, '
                f'{self.type!r})')

    def __str__(
            self,
            *arg,
            **kw
            ) -> str:
        return self.value

    def __len__(
            self
            ) -> int:
        return 1

    def __eq__(
            self,
            other
            ) -> bool:
        return (
            hasattr(other, 'type') and
            hasattr(other, 'value') and
            self.type == other.type and
            self.value == other.value)

    def flatten(
            self,
            *arg,
            **kw):
        return self.value


class Operator:
    """Operator with arity > 0."""

    def __init__(
            self,
            operator:
                str,
            *operands
            ) -> None:
        try:
            operator + 'a'
        except TypeError:
            raise TypeError(
                'operator must be string, '
                f'got: {operator}')
        self.type = 'operator'
        self.operator = operator
        self.operands = list(operands)

    def __repr__(
            self
            ) -> str:
        class_name = type(self).__name__
        xyz = ', '.join(map(repr, self.operands))
        return (
            f'{class_name}('
                f'{self.operator!r}, '
                f'{xyz})')

    def __str__(
            self
            ) -> str:
        xyz = ' '.join(map(str, self.operands))
        return f'({self.operator} {xyz})'

    def __len__(
            self
            ) -> int:
        return 1 + sum(map(len, self.operands))

    def flatten(
            self,
            *arg,
            **kw):
        csv = ', '.join(
            x.flatten(*arg, **kw)
            for x in self.operands)
        return f'( {self.operator} {csv} )'
