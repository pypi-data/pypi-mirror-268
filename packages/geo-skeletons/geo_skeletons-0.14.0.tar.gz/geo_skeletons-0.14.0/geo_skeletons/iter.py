import itertools


class SkeletonIterator:
    def __init__(
        self, dict_of_coords: dict, coords_to_iterate: list[str], skeleton
    ) -> None:

        self.dict_of_coords = dict_of_coords
        self.coords_to_iterate = list(coords_to_iterate)
        # Needed to get the ordering right with itertools
        self.coords_to_iterate.reverse()
        self.skeleton = skeleton
        self._compile_list()

    def __iter__(self):
        return self

    def __next__(self):
        self.ct += 1
        if self.ct < len(self.list_of_skeletons):
            return self.list_of_skeletons[self.ct]
        raise StopIteration

    def __call__(self, coords_to_iterate: list[str]):
        self.coords_to_iterate = list(coords_to_iterate)
        # Needed to get the ordering right with itertools
        self.coords_to_iterate.reverse()
        self._compile_list()
        return self

    def _compile_list(self):
        coord_dict = {}

        for coord in self.coords_to_iterate:
            coord_value = self.dict_of_coords.get(coord)
            if coord_value is None:
                raise KeyError(
                    f"Cannot iterate over coord {coord}, since it does not exist: {self.dict_of_coords.keys()}"
                )
            else:
                coord_dict[coord] = coord_value

        coord_tuples = itertools.product(*[val for __, val in coord_dict.items()])
        list_of_skeletons = []
        for ctuple in coord_tuples:
            kwargs = {}
            for n, val in enumerate(ctuple):
                kwargs[list(coord_dict.keys())[n]] = val
            list_of_skeletons.append(self.skeleton.sel(**kwargs))

        self.list_of_skeletons = list_of_skeletons
        self.ct = -1
