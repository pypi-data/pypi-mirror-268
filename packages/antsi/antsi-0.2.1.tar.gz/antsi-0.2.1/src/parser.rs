use crate::{
    ast::{Token, Tokens},
    error::{Error, Reason},
    lexer::{Lexeme, Lexer, SyntaxKind},
};
use std::iter::Peekable;

mod content;
mod markup;
mod style;
mod text;

/// Convert a piece of text, potentially containing styled markup, to a sequence of tokens
pub struct Parser<'source> {
    lexer: Peekable<Lexer<'source>>,
    errors: Vec<Error>,
}

impl<'source> Parser<'source> {
    pub fn new(input: &'source str) -> Self {
        Self {
            lexer: Lexer::new(input).peekable(),
            errors: Vec::new(),
        }
    }

    /// Perform the parsing operation
    pub fn parse(mut self) -> (Vec<Token>, Vec<Error>) {
        let mut tokens = Tokens::default();

        loop {
            tokens.extend(text::text(&mut self).unwrap_or_default());

            if let Some(lexeme) = self.peek() {
                match lexeme {
                    SyntaxKind::ParenthesisOpen => {
                        self.error(Reason::UnescapedControlCharacter('('))
                    }
                    SyntaxKind::ParenthesisClose => {
                        self.error(Reason::UnescapedControlCharacter(')'))
                    }
                    SyntaxKind::SquareBracketOpen => {
                        self.error(Reason::UnescapedControlCharacter('['))
                    }
                    SyntaxKind::SquareBracketClose => {
                        self.error(Reason::UnescapedControlCharacter(']'))
                    }
                    _ => self.error(Reason::Expected(vec![SyntaxKind::Eof])),
                }

                self.bump();
            } else {
                break;
            }
        }

        (tokens.into(), self.errors)
    }

    /// Get the next syntax item from the lexer without consuming it
    pub(crate) fn peek(&mut self) -> Option<SyntaxKind> {
        self.lexer.peek().map(|lexeme| lexeme.kind)
    }

    /// Get the next lexeme from the lexer without consuming it
    pub(crate) fn peek_lexeme(&mut self) -> Option<&Lexeme<'_>> {
        self.lexer.peek()
    }

    /// Pop the next syntax item from the lexer
    pub(crate) fn bump(&mut self) -> Lexeme {
        self.lexer.next().expect("missing token")
    }

    /// Check if the parser is currently at the given syntax item
    pub(crate) fn at(&mut self, kind: SyntaxKind) -> bool {
        self.peek() == Some(kind)
    }

    /// Expect a syntax item, emitting an error if it isn't present
    pub(crate) fn expect(&mut self, kind: SyntaxKind) -> Option<Lexeme> {
        if self.at(kind) {
            Some(self.bump())
        } else {
            self.error(Reason::Expected(vec![kind]));
            None
        }
    }

    /// Consume lexemes until a non-whitespace lexeme is reached
    pub(crate) fn consume_whitespace(&mut self) {
        while let Some(SyntaxKind::Whitespace) = self.peek() {
            self.bump();
        }
    }

    /// Report an error during parsing
    pub(crate) fn error(&mut self, reason: Reason) {
        let (span, at) = match self.peek_lexeme() {
            Some(lexeme) => (Some(lexeme.span), lexeme.kind),
            None => (None, SyntaxKind::Eof),
        };

        self.errors.push(Error { span, at, reason })
    }
}

#[cfg(test)]
mod tests {
    use crate::ast::Token;

    macro_rules! with_source {
        (
            $source:literal,
            $( { $( $name:ident => $value:expr ),+ $(,)? } , )?
            |$result:ident, $errors:ident| $actions:expr
        ) => {
            insta::with_settings!({
                description => $source,
                omit_expression => true,
                $( $( $name => $value, )+ )?
            }, {
                let ($result, $errors) = $crate::parser::Parser::new($source).parse();
                $actions;
            })
        };
    }

    macro_rules! assert_snapshot {
        ( { $( $name:ident => $value:expr ),+ $(,)? }, $expr:expr ) => {
            insta::with_settings!({
                $( $name => $value, )+
            }, {
                insta::assert_debug_snapshot!($expr);
            });
        };
        ($expr:expr) => {
            insta::assert_debug_snapshot!($expr);
        }
    }

    #[test]
    fn parse_empty() {
        with_source!("", |result, errors| {
            assert_eq!(result, vec![]);
            assert_eq!(errors, vec![]);
        });
    }

    #[test]
    fn parse_text() {
        with_source!(
            "this some text with wh ite\nspa\tce and numb3r5 and $ymb@l$ and CAPITALS",
            |result, errors| {
                assert_eq!(errors, vec![]);
                assert_snapshot!(result);
            }
        );
    }

    #[test]
    fn lowercase_alphabetic() {
        with_source!("abcdef", |result, errors| {
            assert_eq!(result, vec![Token::Content(String::from("abcdef"))]);
            assert!(errors.is_empty());
        });
    }

    #[test]
    fn uppercase_alphabetic() {
        with_source!("ABCDEF", |result, errors| {
            assert_eq!(result, vec![Token::Content(String::from("ABCDEF"))]);
            assert!(errors.is_empty());
        });
    }

    #[test]
    fn mixed_case_alphabetic() {
        with_source!("aBcDeF", |result, errors| {
            assert_eq!(result, vec![Token::Content(String::from("aBcDeF"))]);
            assert!(errors.is_empty());
        });
    }

    #[test]
    fn numeric() {
        with_source!("123456", |result, errors| {
            assert_eq!(result, vec![Token::Content(String::from("123456"))]);
            assert!(errors.is_empty());
        });
    }

    #[test]
    fn lowercase_alphanumeric() {
        with_source!("abc123", |result, errors| {
            assert_eq!(result, vec![Token::Content(String::from("abc123"))]);
            assert!(errors.is_empty());
        });
    }

    #[test]
    fn uppercase_alphanumeric() {
        with_source!("ABC123", |result, errors| {
            assert_eq!(result, vec![Token::Content(String::from("ABC123"))]);
            assert!(errors.is_empty());
        });
    }

    #[test]
    fn mixed_case_alphanumeric() {
        with_source!("AbCd1234", |result, errors| {
            assert_eq!(result, vec![Token::Content(String::from("AbCd1234"))]);
            assert!(errors.is_empty());
        });
    }

    #[test]
    fn special_characters() {
        with_source!("!@#$%^", |result, errors| {
            assert_eq!(result, vec![Token::Content(String::from("!@#$%^"))]);
            assert!(errors.is_empty());
        });
    }

    #[test]
    fn mixed_characters() {
        with_source!("ABCdef123!@#", |result, errors| {
            assert_eq!(result, vec![Token::Content(String::from("ABCdef123!@#"))]);
            assert!(errors.is_empty());
        });
    }

    #[test]
    fn escaped_characters() {
        with_source!("\\(\\)\\[\\]", |result, errors| {
            assert_eq!(result, vec![Token::Content(String::from("()[]"))]);
            assert!(errors.is_empty());
        });
    }

    #[test]
    fn escaped_whitespace() {
        with_source!("\\ \n\t", |result, errors| {
            assert_eq!(result, vec![]);
            assert!(errors.is_empty());
        });
    }

    #[test]
    fn empty_token() {
        with_source!("[fg:red]()", |result, errors| {
            assert_snapshot!(result);
            assert!(errors.is_empty());
        });
    }

    #[test]
    fn token_with_foreground() {
        with_source!("[fg:red](inner)", |result, errors| {
            assert_snapshot!(result);
            assert!(errors.is_empty());
        });
    }

    #[test]
    fn token_with_background() {
        with_source!("[bg:blue](inner)", |result, errors| {
            assert_snapshot!(result);
            assert!(errors.is_empty());
        });
    }

    #[test]
    fn token_with_single_decoration() {
        with_source!("[deco:dim](inner)", |result, errors| {
            assert_snapshot!(result);
            assert!(errors.is_empty());
        });
    }

    #[test]
    fn token_with_multiple_decorations() {
        with_source!("[deco:dim,italic](inner)", |result, errors| {
            assert_snapshot!(result);
            assert!(errors.is_empty());
        });
    }

    #[test]
    fn token_with_multiple_styles() {
        with_source!(
            "[deco:dim,italic;fg:red;bg:blue](inner)",
            |result, errors| {
                assert_snapshot!(result);
                assert!(errors.is_empty());
            }
        );
    }

    #[test]
    fn token_with_leading_content() {
        with_source!("leading [fg:red](content)", |result, errors| {
            assert_snapshot!(result);
            assert!(errors.is_empty());
        });
    }

    #[test]
    fn token_with_trailing_content() {
        with_source!("[fg:red](content) trailing", |result, errors| {
            assert_snapshot!(result);
            assert!(errors.is_empty());
        });
    }

    #[test]
    fn token_with_leading_and_trailing_content() {
        with_source!("leading [fg:red](content) trailing", |result, errors| {
            assert_snapshot!(result);
            assert!(errors.is_empty());
        });
    }

    #[test]
    fn nested_token() {
        with_source!("[fg:red]([bg:blue](inner))", |result, errors| {
            assert_snapshot!(result);
            assert!(errors.is_empty());
        });
    }

    #[test]
    fn kitchen_sink() {
        with_source!(
            "leading [fg:red](one [bg:blue](two [deco:dim](three) two) one) trailing",
            |result, errors| {
                assert_snapshot!(result);
                assert!(errors.is_empty());
            }
        );
    }

    #[test]
    fn parse_unescaped_open_parenthesis_in_plaintext() {
        with_source!("before ( after", |result, errors| {
            assert_snapshot!({ snapshot_suffix => "result" }, result);
            assert_snapshot!({ snapshot_suffix => "errors" }, errors);
        });
    }

    #[test]
    fn parse_unescaped_close_parenthesis_in_plaintext() {
        with_source!("before ) after", |result, errors| {
            assert_snapshot!({ snapshot_suffix => "result" }, result);
            assert_snapshot!({ snapshot_suffix => "errors" }, errors);
        });
    }

    #[test]
    fn parse_unescaped_open_square_bracket_in_plaintext() {
        with_source!("before [ after", |result, errors| {
            assert_eq!(result, vec![]);
            assert_snapshot!({ snapshot_suffix => "errors" }, errors);
        });
    }

    #[test]
    fn parse_unescaped_close_square_bracket_in_plaintext() {
        with_source!("before ] after", |result, errors| {
            assert_snapshot!({ snapshot_suffix => "result" }, result);
            assert_snapshot!({ snapshot_suffix => "errors" }, errors);
        });
    }

    #[test]
    fn parse_unescaped_open_parenthesis_in_token() {
        with_source!("[fg:red](before ( after)", |result, errors| {
            assert_snapshot!({ snapshot_suffix => "result" }, result);
            assert_snapshot!({ snapshot_suffix => "errors" }, errors);
        });
    }

    #[test]
    fn parse_unescaped_close_parenthesis_in_token() {
        with_source!("[fg:red](before ) after)", |result, errors| {
            assert_snapshot!({ snapshot_suffix => "result" }, result);
            assert_snapshot!({ snapshot_suffix => "errors" }, errors);
        });
    }

    #[test]
    fn parse_unescaped_open_square_bracket_in_token() {
        with_source!("[fg:red](before [ after)", |result, errors| {
            assert_eq!(result, vec![]);
            assert_snapshot!({ snapshot_suffix => "errors" }, errors);
        });
    }

    #[test]
    fn parse_unescaped_close_square_bracket_in_token() {
        with_source!("[fg:red](before ] after)", |result, errors| {
            assert_snapshot!({ snapshot_suffix => "result" }, result);
            assert_snapshot!({ snapshot_suffix => "errors" }, errors);
        });
    }

    #[test]
    fn parse_bad_escape_character() {
        with_source!("before \\a after", |result, errors| {
            assert_snapshot!({ snapshot_suffix => "result" }, result);
            assert_snapshot!({ snapshot_suffix => "errors" }, errors);
        });
    }

    #[test]
    fn parse_bad_escape_character_in_token() {
        with_source!("[fg:red](before \\a after)", |result, errors| {
            assert_snapshot!({ snapshot_suffix => "result" }, result);
            assert_snapshot!({ snapshot_suffix => "errors" }, errors);
        });
    }
}
