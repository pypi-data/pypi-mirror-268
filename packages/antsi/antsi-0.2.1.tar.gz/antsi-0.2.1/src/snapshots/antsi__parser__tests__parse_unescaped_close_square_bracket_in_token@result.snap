---
source: src/parser.rs
description: "[fg:red](before ] after)"
---
[
    Content(
        " after",
    ),
]
